midend/checkTableEntries.cpp

```diff
@@ -0,0 +1,106 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "checkTableEntries.h"
+
+using namespace P4::literals;
+
+void P4::CheckTableEntries::get_mask_val(const IR::Expression *e, big_int &mask, big_int &val) {
+    if (auto *m = e->to<IR::Mask>()) {
+        auto *l = m->left->to<IR::Constant>();
+        auto *r = m->right->to<IR::Constant>();
+        if (l && r) {
+            mask = r->value;
+            val = l->value;
+        } else
+            error(ErrorType::ERR_UNEXPECTED, "%1% Unexpected entry key expression", e);
+    } else if (auto *k = e->to<IR::Constant>()) {
+        val = k->value;
+        if (auto w = k->type->width_bits(); w > 0)
+            mask = IR::Constant::GetMask(w).value;
+        else
+            mask = -1;
+    } else if (auto *bl = e->to<IR::BoolLiteral>()) {
+        mask = 1;
+        val = bl->value ? 1 : 0;
+    } else if (e->is<IR::DefaultExpression>()) {
+        mask = val = 0;
+    } else {
+        error(ErrorType::ERR_UNEXPECTED, "%1% Unexpected entry key expression", e);
+    }
+}
+
+bool P4::CheckTableEntries::ternary_covers(const IR::Expression *k1, const IR::Expression *k2) {
```

Comment:
For future readers, a comment like the following on `ternary_covers` might be helpful:
```
// Return true if the set of lookup keys matched by k1 is a superset
// of the keys matched by k2, including if they match the same set of lookup keys.
// Otherwise return false.
```